tortoise/verifying_tortoise.go

```diff
@@ -419,75 +398,65 @@ func (t *turtle) calculateExceptions(
 	if startLayer.Before(types.GetEffectiveGenesis()) {
 		startLayer = types.GetEffectiveGenesis()
 	}
-	for layerID := startLayer; !layerID.After(t.Last); layerID = layerID.Add(1) {
-		logger := logger.WithFields(log.Named("diff_layer_id", layerID))
+	for lid := startLayer; !lid.After(t.Last); lid = lid.Add(1) {
+		logger := logger.WithFields(log.Named("diff_layer_id", lid))
 		logger.Debug("checking input vector diffs")
 
-		layerBlockIds, err := t.getLayerBlocksIDs(ctx, layerID)
+		local, err := t.getLocalOpinion(ctx, lid)
 		if err != nil {
 			if errors.Is(err, database.ErrNotFound) {
 				continue
 			}
-			return nil, fmt.Errorf("layer block IDs: %w", err)
+			return nil, err
 		}
 
 		// helper function for adding diffs
-		addDiffs := func(bid types.BlockID, voteClass string, voteVec vec, diffMap map[types.BlockID]struct{}) {
-			if v, ok := baseBlockOpinion[bid]; !ok || simplifyVote(v) != voteVec {
-				logger.With().Debug("added vote diff",
-					log.Named("diff_block", bid),
-					log.String("diff_class", voteClass),
-					log.Named("opinion", v),
-				)
-				if layerID.Before(baseBlockLayerID) {
-					logger.With().Warning("added exception before base block layer, this block will not be marked good",
-						log.Named("diff_block", bid),
-						log.String("diff_class", voteClass))
+		addDiffs := func(logger log.Log, bid types.BlockID, voteVec vec, diffMap map[types.BlockID]struct{}) {
+			if v, ok := baseopinion[bid]; !ok || simplifyVote(v) != voteVec {
+				logger.With().Debug("added vote diff", log.Named("opinion", v))
+				if lid.Before(baselid) {
+					logger.With().Warning("added exception before base block layer, this block will not be marked good")
 				}
 				diffMap[bid] = struct{}{}
 			}
 		}
 
-		// get local opinion for layer
-		layerInputVector, err := t.layerOpinionVector(ctx, layerID)
-		if err != nil {
-			// an error here signifies a real database failure
-			logger.With().Error(errstrUnableToCalculateLocalOpinion, log.Err(err))
-			return nil, err
-		}
-
-		// otherwise, nil means we should abstain
-		if layerInputVector == nil {
-			// still waiting for Hare results, vote neutral and move on
-			logger.With().Debug("input vector is empty, adding neutral diffs", log.Err(err))
-			for _, b := range layerBlockIds {
-				addDiffs(b, "neutral", abstain, neutralDiff)
+		for bid, opinion := range local {
+			usecoinflip := opinion == abstain && lid.Before(t.layerCutoff())
+			if usecoinflip {
+				coin, exist := t.bdp.GetCoinflip(ctx, votinglid)
+				if !exist {
+					return nil, fmt.Errorf("coinflip is not recorded in %s", votinglid.Sub(1))
```

Comment:
this would happen when the node didn't participate in the hare process for that votinglib.Sub(1) layer.
i think not-participating in hare is common. so it come downs to how often do we need to use coin flip. i think we should at least open an issue for the missing coin flip. do you think we can record it in the hare certificate? or did we already?

Reply:
> i think not-participating in hare is common. so it come downs to how often do we need to use coin flip. 

there is some misunderstanding. i think that participating in hare voting is not a prerequisite for receiving a coin, i will doublecheck.